### ISSUE TITLE
Move server settings to somewhere

### ISSUE THREAD

**Issue by rwols** [sublimelsp/LSP#1048]
The current way of specifying settings is awkward. You have to edit LSP.sublime-settings to modify a server setting. The LSP.sublime-settings file is more suited for general server startup info and language assignments.

I propose we move the (dotted #1047) settings to Preferences.sublime-settings. This has multiple advantages:

- The `settings` dicts in LSP.sublime-settings play the role as defaults and which keys to lookup. You won't have to touch them often.
- The dotted keys are assigned to the `view` automatically. So fetching settings becomes as easy as `view.settings().get(dotted_key)`.
- Overrides from .sublime-project files are automatic, no extra code needed.

**Comment by rwols** [comment]
Blocked by #1047 for now.

**Comment by rchl** [comment]
It's very standard practice for packages to have their own namespaced settings.
I wouldn't really like to mix those with global (`Preferences.sublime-settings`) settings.

The main reason for me why I like it the way it is now is a clear separation of plugin and global settings.

Even though your suggested change would make overrides simpler (from the code's perspective), it's still possible to do overrides while having namespaced settings (and also clearer when overrides in project files follow convention like convention like `settings: { lsp.some_setting: true }`.

**Comment by rwols** [comment]
I'm sorry for being unclear. I'm talking about *server settings*.

So for instance:
```js
// LSP.sublime-settings
{
    "clients":
    {
        "pyls":
        {
            "settings":
            {
                // Sane defaults
                "pyls.plugins.pydocstyle.enabled": false
            }
        }
    }
}

// Preferences.sublime-settings
{
    "pyls.plugins.pydocstyle.enabled": true
}
```

> Also, how would you name LSP settings then?

It's best if they stay in LSP.sublime-settings.

**Comment by rwols** [comment]
Well, I suppose they can also be prefixed with `lsp.`, but I think that's again a bit unclear for end-users when they're reading READMEs of servers.

**Comment by rchl** [comment]
Can't we just do:
```js
// LSP.sublime-settings
{
  "pyls.plugins.pydocstyle.enabled": true
}
```
?

The advantage being, that the server-specific settings are scoped to LSP which is where they belong IMO. We don't need to (and shouldn't IMO) fully base our behavior on vscode as it has no concept of scoped settings. It puts everything in the same place.

As for the project overrides, it could still look like this, I believe:
```
// MyProject.sublime-project
{
  "folders": [],
  "settings":
  {
    "pyls.plugins.pydocstyle.enabled": true
  }
}
```
But could instead be (slightly more difficult to understand for the user maybe):
```
// MyProject.sublime-project
{
  "folders": [],
  "settings":
  {
    "lap"
        "pyls.plugins.pydocstyle.enabled": true
    }
  }
}
```

**Comment by rwols** [comment]
You may be right.

Let's reason backwards from `workspace/didChangeConfiguration`.

We want to send this notification if a setting from a server is modified.
The ST API provides a `sublime.Settings` object. We can attach a "changed" callback to it via `add_on_change` and remove the attached callback with `clear_on_change`. The callback is invoked once *something* in the settings object changes. The callback does not receive which key changed.

So we want to use this sublime.Settings object to our advantage: send workspace/didChangeConfiguration with all server settings once something changes in that sublime.Settings object.

Each sublime.View has its own personal sublime.Settings object, so this does not mesh well with workspace/didChangeConfiguration (if there were a textDocument/didChangeConfiguration, then it would work well). So putting the settings in Preferences.sublime-settings is not a good idea.

Each .sublime-settings file on-disk also has an associated sublime.Settings object. If it's in a Package/xxx location, and there's a .sublime-settings file with the same name in Packages/User, then those are merged into the same sublime.Settings object.

If we put the server settings in LSP.sublime-settings, then what can happen is:

- User modified startup info of a server (e.g. `"command"` or `"env"`), or
- User modified a server setting.

We would have to detect this if we want to cleanly notify workspace/didChangeConfiguration. We can also be blunt and restart all servers once something in LSP.sublime-settings changes.

We can also add a new, separate LSP-server-settings.sublime-settings file specifically for the server settings. We can then have a separate callback and we can be certain that only the server settings were modified. Thus we can send workspace/didChangeConfiguration.

Writing this out I think it'd be best if were to add a new settings file called LSP-server-settings.sublime-settings for workspace/didChangeConfiguration.

We can then remove the server settings in LSP.sublime-settings.

Since we cannot iterate over the key-value pairs of a sublime.Settings object we would have to use `sublime.load_resource` and `sublime.decode_value`.

For the `workspace/configuration` request of a server we *can* use the sublime.Settings object loaded through `sublime.load_settings("LSP-server-settings.sublime-settings")`.

**Comment by rchl** [comment]
I guess for `LSP-*` packages the problem is more trivial to handle since there is individual settings file for each.

But for:
> Writing this out I think it'd be best if we were to add a new settings file called LSP-server-settings.sublime-settings for workspace/didChangeConfiguration.

This is a bit controversial I guess. Another settings file will probably not make things clearer to the users. But I'm not against if you think that's the way to go.

I wonder if there is anything that could be done in ST API to help that specific case...

**Comment by rwols** [comment]
I think if we add a new command `LSP: Edit Server Settings` in the command palette it should be discoverable enough...

**Comment by rchl** [comment]
Sounds good.
And let's include migration code for that so that user doesn't have to do anything.

**Comment by AmjadHD** [comment]
how are you planning to handle changes in project settings ? since `on_change` event is not triggered for project settings.

**Comment by AmjadHD** [comment]
how are you planning to handle changes in project settings ? since `on_change` event is not triggered for project settings.

**Comment by deathaxe** [comment]
> It's very standard practice for packages to have their own namespaced settings.
> I wouldn't really like to mix those with global (_Preferences.sublime-settings_) settings.

To add my 5ct about it:

1. Yes, this is the main way plugin settings are handled.
2. It works well as long as all plugin settings are globally applied and/or kind of static.
3. As soon as view/syntax/project specific settings are to be supported the best/simplest option seems to be to place everything into _Preferences.sublime-settings_ to make use of ST's preferences-overlay feature. GitGutter's settings classes were extended this way to support all kinds of settings in both, the GitGutter.sublime-settings (for legacy reasons) and Preferences.sublime-settings. This is the way how to enable/disable GitGutter per project, per syntax or per view.
4. If LSP settings are to be located in a plugin specific LSP.sublime-settings, they should not be placed into the `"settings"` dict of a _.sublime-project_ file, but into a dedicated "LSP" dict, then. It just feels odd to partly hijack sublime settings on a per-project level (which also requires some kind of namespacing via dotted settings).
5. PackageDev already introduced syntax support for namespaced/dotted settings like `"packagedev.thesetting"`. The main intention was to use the package name as main/first part of the name separated by dots. It would mean to use `"lsp.servers.pyls.enabled"` or `"lsp.servers.pyls.plugins.pycodestyle.enabled"` etc. . ALL settings should start with `"lsp."`.
6. Placing everything as dotted settings into _Preferences_ would also solve any kind of issues related with merging user defined settings with defaults.
7. Dotted settings would be the easiest way for PackageDev or jsonschema to provide documentation and autocompletion.
8. The current way of caching settings and try to keep them up to date would be obsolete when using _Preferences.sublime-settings_ as well. Just read the needed setting when needed would always return the up to date value for a view, no matter where changes have been made by a user.

SublimeLinter used to manage the same kind of nested settings dicts as LSP does at the moment. They quickly realized it's not manageable. The issues outlined with all the LSP sub-plugins are only one part of all that stuff.

**Comment by rwols** [comment]
> PackageDev already introduced syntax support for namespaced/dotted settings like "packagedev.thesetting". The main intention was to use the package name as main/first part of the name separated by dots. It would mean to use "lsp.servers.pyls.enabled" or "lsp.servers.pyls.plugins.pycodestyle.enabled" etc. . ALL settings should start with "lsp."

The problem with this is is that whenever some other unrelated settings change, e.g. `font_size`, we get a change callback from the Settings object. Should we send `workspace/didChangeConfiguration` in any case?

**Comment by rwols** [comment]
What's also problematic is that `workspace/didChangeConfiguration` is a workspace (LSP) / window (ST) level concept, but `view.settings()` are a textDocument (LSP) / view (ST) concept.

So if I have two views, one with setting `"lsp.foo.bar": 1` and one with setting "`lsp.foo.bar": 2`, I don't know which value to send in `workspace/didChangeConfiguration`.

**Comment by rwols** [comment]
It is true that for the `workspace/configuration` *request* from a server, we *can* use `view.settings()` if the server provides a `scopeUri` that matches with some view's filename.

OK, so I suppose we can use `sublime.load_settings("Preferences.sublime-settings")` for `workspace/didChangeConfiguration`, and when a server asks for more specific settings, we can use `view.settings()` if the `scopeUri` matches.

But all this does mean we'll still be stuck with a `"settings"` dict in our server configurations, because https://github.com/sublimehq/sublime_text/issues/2916

**Comment by deathaxe** [comment]
Just providing some infos about experiences I've made. I am far away from being convinced placing everything into _Preferences_ is the best solution for LSP. It just worked well for GitGutter's purpose. 

> Should we send `workspace/didChangeConfiguration` in any case?

With a look at the lemminx code the message is just decoded and the values are applied to the server's internal configuration object. We just might send too many information then.

If sending unchanged configurations is unwanted, we'd still need to cache and compare relevant settings with the previous state (which was sent).

> `workspace/didChangeConfiguration` is a workspace (LSP) / window (ST) level concept, ...

Unfortunatelly there's no real good concept for project level settings without view specific settings included.

I just have the discussion about indentation settings (https://github.com/sublimelsp/LSP-lemminx/issues/1) in mind. As per indentation detection each file can have different settings, which would need to be sent to the server in some way. It is a per file setting.

Even if we register a view-level settings changed callback, we could still try to read project data and syntax specfici or user preferences if in doupt whether view specific settings might break things.

Not quite sure how a good solution would look like though.

**Comment by rwols** [comment]
It's been decided that server settings should stay near the server configuration. It says in `"settings"`. There's special code to merge  overrides from project files and User/foo.sublime-settings files.
